### PR TITLE
Pinned sidebar

### PR DESCRIPTION
**Pull request by MattBudz**
### Summary

This PR allows the sidebar to remain expanded or collapsed across views. Without forcing the user to collapse the sidebar in order to use the app.

### Check List

- [x] Added a CHANGELOG entry

### CODE REVIEW

File: app/assets/stylesheets/tylium/modules/_secondary_sidebar.scss [hunk 1]
```diff
@@ -2,6 +2,12 @@
   background: $sidebarActiveColor;
   color: $lightColor;
   border-right: 1px solid $borderColor;
+  padding: 1rem;
```
Comment (brianp) [inline app/assets/stylesheets/tylium/modules/_secondary_sidebar.scss]: Alpha.

File: app/assets/stylesheets/tylium/modules/_secondary_sidebar.scss [hunk 2]
```diff
@@ -11,146 +17,135 @@
...
+      margin-top: 0.5rem;
+      cursor: default;
+    
+      a {
+        color: $lightColor;
       }
 
-      .header-underline {
-        border-color: $lightColor;
+      &:first-of-type{
```
Comment (brianp) [inline app/assets/stylesheets/tylium/modules/_secondary_sidebar.scss]:
```suggestion
      &:first-of-type {
```

File: app/views/layouts/tylium/_sidebar.html.erb [hunk 3]
```diff
@@ -1,4 +1,4 @@
-<div class="main-sidebar sidebar-collapsed" data-behavior="main-sidebar sidebar-collapsed">
```
Comment (brianp) [inline app/views/layouts/tylium/_sidebar.html.erb]: TIL. I didn't realize you could add multiple values to a data behaviour. I assumed it would create a single behaviour named the full string "main-sidebar sidebar-collapsed".

File: app/assets/javascripts/tylium/behaviors.js.coffee [hunk 4]
```diff
@@ -217,3 +221,8 @@ document.addEventListener "turbolinks:load", ->
   $('[data-behavior~=smooth-scroll]').on 'click', ->
     target = $(this).data('target');
     $('[data-id~=' + target + ']')[0].scrollIntoView({behavior: "smooth"});
+
+document.addEventListener "turbolinks:before-cache", ->
```
Comment (brianp) [inline app/assets/javascripts/tylium/behaviors.js.coffee]: I understand why we need this. I don't like it one bit, but I understand why.

Feels like we're often fighting turbolinks. I think the optimum solution would be for us to get the sidebar outside of tubolinks viewport entirely. That is outside the scope of this issue though.

File: app/assets/javascripts/tylium/behaviors.js.coffee [hunk 5]
```diff
@@ -160,33 +160,37 @@ document.addEventListener "turbolinks:load", ->
...
+    localStorage.setItem('sidebar-view-state', $sidebar.attr('data-state'));
+
+  sidebarClose = (animationClass) ->
+    $navbar.css('left', '0px')
+    $sidebar.removeClass('sidebar-expanded no-animation animate').addClass('sidebar-collapsed' + animationClass).attr('data-state', 'sidebar-collapsed')
+    $viewContent.css({'left': '43px', 'width': 'calc(100vw - 43px)'})
+    # This is required to keep the sidebar state if the user refreshes the page
+    localStorage.setItem('sidebar-view-state', $sidebar.attr('data-state')); 
+
+  if localStorage.getItem('sidebar-view-state') && localStorage.getItem('sidebar-view-state') == 'sidebar-collapsed'
```
Comment (brianp) [inline app/assets/javascripts/tylium/behaviors.js.coffee]:
```suggestion
  if localStorage.getItem('sidebar-view-state') == 'sidebar-collapsed'
```

You don't need to see if the value exists first. It will return `null` if it does not, and `null` == 'sidebar-collapsed' will be `false`

File: app/assets/stylesheets/tylium/modules/_secondary_sidebar.scss [hunk 6]
```diff
@@ -2,6 +2,12 @@
   background: $sidebarActiveColor;
   color: $lightColor;
   border-right: 1px solid $borderColor;
+  padding: 1rem;
+  height: calc(100vh - 60px);
+  overflow: auto;
+  z-index: 2;
+  position: sticky;
+  top: 0;
```
Comment (MattBudz) [inline app/assets/stylesheets/tylium/modules/_secondary_sidebar.scss]:
```suggestion
  background: $sidebarActiveColor;
  border-right: 1px solid $borderColor;
  color: $lightColor;
  height: calc(100vh - 60px);
  overflow: auto;
  padding: 1rem;
  position: sticky;
  top: 0;
  z-index: 2;
```

File: app/views/layouts/tylium/_sidebar.html.erb (hunk 3 above)
Comment (MattBudz) [reply app/views/layouts/tylium/_sidebar.html.erb]: @brianp yup, we can stack them in the erb's and then use `~=` in JS when selecting elements by data-attribute.

File: app/assets/javascripts/tylium/behaviors.js.coffee (hunk 4 above)
Comment (etdsoft) [reply app/assets/javascripts/tylium/behaviors.js.coffee]: shouldn't you just save the state as part of the `click` handler?

File: app/assets/javascripts/tylium/behaviors.js.coffee [hunk 7]
```diff
@@ -160,33 +160,37 @@ document.addEventListener "turbolinks:load", ->
...
+  if localStorage.getItem('sidebar-view-state') == 'sidebar-collapsed'
+    sidebarClose(' no-animation')
+  else
+    sidebarOpen(' no-animation')
 
   $('[data-behavior~=sidebar-toggle]').on 'click', ->
-    if $sidebar.is('[data-behavior~=sidebar-collapsed]')
-      sidebarOpen()
+    if $sidebar.is('[data-state~=sidebar-collapsed]')
+      sidebarOpen(' animate')
```
Comment (etdsoft) [inline app/assets/javascripts/tylium/behaviors.js.coffee]: Maybe add the extra space to L176 instead of having to remember to pass it in the argument.

File: app/assets/javascripts/tylium/behaviors.js.coffee [hunk 8]
```diff
@@ -160,33 +160,37 @@ document.addEventListener "turbolinks:load", ->
...
       else
-        sidebarClose()
-
-  $('[data-behavior~=back-fade]').on 'click', ->
-    sidebarClose()
-
-  $('[data-behavior~=sidebar-link]').on 'click', ->
-    if $sidebar.is('[data-behavior~=sidebar-expanded]')
-      sidebarClose() 
+        sidebarClose(' animate')
```
Comment (etdsoft) [inline app/assets/javascripts/tylium/behaviors.js.coffee]: ditto
